File: python/singleton.py
# This example demonstrates singleton design pattern in Python.

class Singleton:
    __instance = None

    @staticmethod 
    def get_instance():
        if Singleton.__instance == None:
            Singleton()
        return Singleton.__instance

    def __init__(self):
        if Singleton.__instance != None:
            raise Exception("This class is a singleton!")
        else:
            Singleton.__instance = self

a = Singleton()
print(a)

b = Singleton.get_instance()
print(b)

c = Singleton.get_instance()
print(c)

# Output:
# <__main__.Singleton object at 0x7f9d38033bb0>
# <__main__.Singleton object at 0x7f9d38033bb0>
# <__main__.Singleton object at 0x7f9d38033bb0>

# The Singleton class has a private static attribute __instance.
# The get_instance() method returns the __instance attribute.
# The __init__() method is called only once when the __instance attribute is None.
# This is called singleton design pattern.

File: javascript/singleton.js
// This example demostrate the singleton in javascript
// The singleton is a design pattern that restricts the instantiation of a class to one object.

// Example 1
// The following example demostrate the singleton
var singleton = (function() {
    var instance;
    function init() {
        var name = "John";
        function getName() {
            return name;
        }
        return {
            getName: getName
        }
    }
    return {
        getInstance: function() {
            if (!instance) {
                instance = init();
            }
            return instance;
        }
    }
})();

var singleton1 = singleton.getInstance();
var singleton2 = singleton.getInstance();

console.log(singleton1.getName());
console.log(singleton2.getName());

// Output:
// John
// John